src/identity.js

```diff
@@ -436,93 +436,77 @@ export class Identity extends EventEmitter {
      * @return {Promise<Identity#HasSessionSuccessResponse|Identity#HasSessionFailureResponse>}
      */
     hasSession(autologin = true) {
-        if (this._hasSessionInProgress) {
-            return this._hasSessionInProgress;
-        }
-        const promiseFn = async (resolve, reject) => {
-            const postProcess = (sessionData) => {
+        if (!this._hasSessionInProgress) {
+            if (typeof autologin !== 'boolean') {
+                const [type, value] = inspect(autologin);
+                return Promise.reject(new SDKError(`Parameter 'autologin' must be boolean, was: "${type}:${value}"`));
+            }
+            const _postProcess = (sessionData) => {
                 if (sessionData.error) {
-                    this.emit('error', sessionData.error);
-                    return reject(new SDKError('HasSession failed', sessionData.error));
+                    throw new SDKError('HasSession failed', sessionData.error);
                 }
                 this._maybeSetVarnishCookie(sessionData);
                 this._emitSessionEvent(this._session, sessionData);
             };
-
-            if (typeof autologin !== 'boolean') {
-                const [type, value] = inspect(autologin);
-                return reject(new SDKError(`Parameter 'autologin' must be boolean, was: "${type}:${value}"`));
-            }
-
-            try {
+            const _getSession = async () => {
+                let sessionData = null;
                 if (this._enableSessionCaching) {
                     // Try to resolve from cache (it has a TTL)
-                    const cachedData = this.cache.get(HAS_SESSION_CACHE_KEY);
-                    if (cachedData) {
-                        postProcess(cachedData);
-                        return resolve(cachedData);
-                    }
```

Comment:
This PR centralizes the call to `_postProcess` in a single place, after all different ways of obtaining the session has been tried. Some additional guards are added to not clobber previously set `sessionData`